### PR TITLE
Implementation OOB and disOOB strategies from byeDPI

### PR DESCRIPTION
**Pull request by siriusfreak**
This PR contains the implementation OOB and disOOB strategies from https://github.com/hufrea/byedpi/blob/main/desync.c

### CODE REVIEW

File: x/oob/writer.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	delay time.Duration,
+) io.Writer {
+	return &oobWriter{conn: conn, opts: opts, oobPosition: oobPosition, oobByte: oobByte, disOOB: disOOB, delay: delay}
+}
+
+func (w *oobWriter) Write(data []byte) (int, error) {
+	var written int
+	var err error
+
+	if w.oobPosition > 0 && w.oobPosition < int64(len(data))-1 {
```
Comment (fortuna) [inline x/oob/writer.go:50]: There's a corner case here. Probably worth capturing in the tests.

If the oob position aligns with the end of a write, it will never happen.

The problem is that you are dependent on an extra byte from the following write to push the OOB byte.

There's a way to fix that without requiring memory allocation. The SendMsg call can take the OOB as a separate buffer!

On Windows, the [WSASend](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsasend) call takes a list of buffers, so just append the OOB to that list.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: x/oob/writer.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	opts sockopt.TCPOptions,
+	oobPosition int64,
+	oobByte byte,
+	disOOB bool,
+	delay time.Duration,
+) io.Writer {
+	return &oobWriter{conn: conn, opts: opts, oobPosition: oobPosition, oobByte: oobByte, disOOB: disOOB, delay: delay}
+}
+
+func (w *oobWriter) Write(data []byte) (int, error) {
```
Comment (PeterZhizhin) [inline x/oob/writer.go:46]:
```
func (w *oobWriter) Write(data []byte) (written int, err error) {
```

File: x/configurl/oob.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+func registerOOBStreamDialer(r TypeRegistry[transport.StreamDialer], typeID string, newSD BuildFunc[transport.StreamDialer]) {
+	r.RegisterType(typeID, func(ctx context.Context, config *Config) (transport.StreamDialer, error) {
+		sd, err := newSD(ctx, config.BaseConfig)
+		if err != nil {
+			return nil, err
+		}
+		params := config.URL.Opaque
+
+		splitStr := strings.Split(params, ":")
```
Comment (PeterZhizhin) [inline x/configurl/oob.go:22]: nit: maybe named arguments? https://github.com/Jigsaw-Code/outline-sdk/pull/324/files#diff-974d5fac785c258e952052dd1f6aa5b012bf2e4e43eb708de99420175a67e99b

But it's a matter of preference. @fortuna should know better.

Also, maybe you can make most arguments optional? I believe most users can use default timeout, default URG byte value, etc.

File: x/oob/writer.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+func (w *oobWriter) send(data []byte, flags int) error {
+	// Use SyscallConn to access the underlying file descriptor safely
+	rawConn, err := w.conn.SyscallConn()
+	if err != nil {
+		return fmt.Errorf("oob strategy was unable to get raw conn: %w", err)
+	}
+
+	// Use Control to execute Sendto on the file descriptor
+	var sendErr error
+	err = rawConn.Control(func(fd uintptr) {
```
Comment (PeterZhizhin) [inline x/oob/writer.go:108]: nit: semantically it's a `Write` rather than `Control`.  Feel free to keep as is, if you feel like it.

File: x/oob/writer.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		secondPart := data[w.oobPosition:]
+
+		// Split the data into two parts
+		tmp := secondPart[0]
+		secondPart[0] = w.oobByte
+
+		var oldTTL int
+		if w.disOOB {
+			w.setTTL.Do(func() {
+				oldTTL, err = w.opts.HopLimit()
```
Comment (PeterZhizhin) [inline x/oob/writer.go:61]: Take a look at the suggestion in https://github.com/Jigsaw-Code/outline-sdk/pull/329/files#r1840884687 that could simplify this and reduce code duplication.

**Review by PeterZhizhin — COMMENTED** [review]
Amazing, thank you!
